File: BOJ/[10844] 쉬운 계단 수/ONE/10844.py
num = int(input())
## row: n번째 계단, col: col의 숫자로 시작
## 2차원 배열 생성
stair_num = [[0 for _ in range(10)] for _ in range(num + 1)]

for i in range(1, num + 1): # 1 부터 N 계단까지
    for j in range(10): # 0 ~ 9까지
        if i == 1:
            stair_num[i][j] = 1
        else:
            if j == 0: # 0으로 시작 -> N-1번째의 1로 시작하는 계단수
                stair_num[i][j] = stair_num[i-1][1]
            elif j == 9: # 9로 시작 -> N-1번째의 8로 시작하는 계단수
                stair_num[i][j] = stair_num[i-1][8]
            else: # 1~8로 시작 -> N-1번째의 j-1과 j+1 을 더한값
                stair_num[i][j] = stair_num[i-1][j-1] + stair_num[i-1][j+1]
print(sum(stair_num[num][1:10]) % 1000000000)

File: BOJ/[10844] 쉬운 계단 수/Baeksu/10844.cpp
#include <iostream>

using namespace std;

int N;
int dp[100][10];

void dp_stair(int num);

int main() {
	cin >> N;
	for (int i = 0; i < 10; i++) {
		dp[0][i] = 1;
	}

	dp_stair(N);
	return 0;
}

void dp_stair(int num) {
	long long sum = 0;
	for (int i = 1; i < N; i++) {
		dp[i][9] = dp[i - 1][8] % 1000000000;
		dp[i][0] = dp[i][9];
		for (int j = 8; j > 0; j--) {
			dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j + 1];
			dp[i][j] %= 1000000000;
		}
	}

	for (int i = 1; i < 10; i++) {
		sum += dp[num - 1][i];
	}

	cout << sum % 1000000000;
}